File: API_Examples/change_sis_ids/change_user_sisid.py
#!/usr/bin/env python

import requests, json
import csv

""" Example change_users.csv file

old_user_id,new_user_id
201.bobby,2013.bobby

"""

filename = 'change_users.csv' # Change this to the file with the two columns
domain = '<some_domain>.instructure.com' # Change this
token = '<your_token_here>' # Change this


 
# Read CSV file

headers = {'Authorization':'Bearer %s' % token}
user_list = csv.DictReader(open(filename,'rb'))
for user in user_list:
  # Get the logins for the user, find the one with the old SIS id, and change it
  url = 'https://%s/api/v1/users/sis_user_id:%s/logins' % (domain,user['old_user_id'])

  login_response = requests.get(url,headers=headers)
  logins = login_response.json()
  if type(logins) != list:
    print "the user probably wasn't found, keep going"

  else:
    for l in logins:
      try:
        if l['sis_user_id'] == user['old_user_id']:
          params = {'login[sis_user_id]':user['new_user_id']}
          url = 'https://%s/api/v1/accounts/self/logins/%s' % (domain,l['id'])
          updated_login_response = requests.put(url,headers=headers,params=params)
          print updated_login_response.json()
      except Exception, exc:
        print logins
        print "The user probably wasn't found.  Keep going..."

File: API_Examples/change_sis_ids/ruby/change_user_sis_id.rb
require 'unirest'
require 'csv'

#================
# Change these
access_token = nil
domain = nil
env = nil
csv_file = nil

#================
# Don't edit from here down unless you know what you're doing.

unless access_token
  "Puts what is your access token?"
  access_token = gets.chomp
end

unless domain
  "Puts what is your Canvas domain?"
  domain = gets.chomp
end

unless csv_file
  "Puts where is your SIS ID update CSV located?"
  csv_file = gets.chomp
end

unless File.exists?(csv_file)
  raise "Error: can't locate the update CSV"
end

env ? env << "." : env
base_url = "https://#{domain}.#{env}instructure.com/api/v1"
test_url = "#{base_url}/accounts/self"

Unirest.default_header("Authorization", "Bearer #{access_token}")

# Make generic API call to test token, domain, and env.
test = Unirest.get(test_url)

unless test.code == 200
  raise "Error: The token, domain, or env variables are not set correctly"
end

CSV.foreach(csv_file, {:headers => true}) do |row|
  url = "#{base_url}/users/sis_user_id:#{row[0]}/logins"
  login_response = Unirest.get(url)
  logins = login_response.body

  unless logins
    puts "User #{row[0]} wasn't found."
    next
  end

  logins.each do |login|
    hashed = login.to_h
    puts hashed['id']
    puts row[0]
    if hashed['sis_user_id'] == row[0]
      update_id = Unirest.put("#{base_url}/accounts/self/logins/#{hashed['id']}", parameters: { 'login[sis_user_id]' => row[1] })
      puts "Successfully updated \n #{update_id.body}"
    else
      puts "User #{row[0]} wasn't found... Moving right along."
    end
  end
end

puts "Finished updating user SIS IDs."
